File: 0528-random-pick-with-weight/Solution.cs
public class Solution {
    private int[] _weights;

    public Solution(int[] w)
    {
        _weights = new int[w.Length];
        var max = 0;
        for (var i = 0; i < w.Length; i++)
        {
            max += w[i];
            _weights[i] = max;
        }
    }
    
    public int PickIndex() {
        var random = new Random();
        var target = random.Next(0, _weights[_weights.Length - 1]);
        int left = 0, right = _weights.Length - 1;
        while (left < right)
        {
            int mid = left + (right - left) / 2;
            if (_weights[mid] <= target)
                left = mid + 1;
            else
                right = mid;
        }
        return left;
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(w);
 * int param_1 = obj.PickIndex();
 */

File: 0528-random-pick-with-weight/Solution.go
package Leetcode

import "math/rand"

type Solution struct {
	_weights []int
}

func Constructor(w []int) Solution {
	_weights := make([]int, len(w))
	_weights[0] = w[0]
	for i := 1; i < len(w); i++ {
		_weights[i] = _weights[i-1] + w[i]
	}
	return Solution{_weights}
}

func (this *Solution) PickIndex() int {
	randNum := rand.Intn(this._weights[len(this._weights)-1])
	left, right := 0, len(this._weights)-1
	for left < right {
		mid := left + (right-left)/2
		if randNum < this._weights[mid] {
			right = mid
		} else {
			left = mid + 1
		}
	}
	return left
}

/**
 * Your Solution object will be instantiated and called as such:
 * obj := Constructor(w);
 * param_1 := obj.PickIndex();
 */
